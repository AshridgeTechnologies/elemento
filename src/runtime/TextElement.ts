import React from 'react'
import {Typography} from '@mui/material'

export default function TextElement({children}: {children: any}) {
    return React.createElement(Typography, {gutterBottom: true}, children)
}
