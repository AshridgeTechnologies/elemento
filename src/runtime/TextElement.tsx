import React from 'react'
import {Typography} from '@mui/material'
import {flatten} from 'ramda'

type Properties = {path: string, children?: any}

function asText(content: any) {
    if (React.isValidElement(content)) return content
    const contentValue = typeof content === 'object' && 'value' in content ? content.value : content
    if (typeof contentValue === 'string') {
        if (contentValue.includes('\n')) {
            return flatten(contentValue.split('\n').map((line: string, index: number) => [line, React.createElement('br', {key: index})]))
        }
        return contentValue

    }
    return contentValue
}
export default function TextElement({children, path}: Properties) {
    return React.createElement(Typography, {id: path, gutterBottom: true}, asText(children))
}
