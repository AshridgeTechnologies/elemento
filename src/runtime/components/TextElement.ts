import React from 'react'
import {Typography} from '@mui/material'
import lodash from 'lodash'; const {isArray, isFunction, isPlainObject, isObject, isString} = lodash;
import {PropVal, StylesProps, valueLiteral, valueOfProps} from '../runtimeFunctions'
import {flatten} from 'ramda'

type Properties = Readonly<{path: string, children?: any,
    show?: PropVal<boolean>,
    styles?: StylesProps
    }>

function asText(content: any) {
    if (React.isValidElement(content)) return content
    const contentValue = content?.valueOf()
    if (isString(contentValue) && contentValue.includes('\n')) {
        return flatten(contentValue.split('\n').map((line: string, index: number) => [line, React.createElement('br', {key: index})]))
    }
    if (isPlainObject(contentValue)) {
        return valueLiteral(contentValue)
    }
    if (isFunction(contentValue)) {
        return 'function ' + contentValue.name
    }
    if (isObject(contentValue)) {
        return contentValue.toString()
    }
    return contentValue
}

export default function TextElement({children, path, ...props}: Properties) {
    const childArray = isArray(children) ? children : [children]
    const {show = true, styles = {}} = valueOfProps(props)
    const showProps = show ? {} : {display: 'none'}
    const sxProps = {
        ...showProps,
        ...styles
    }

    return React.createElement(Typography, {id: path, sx: sxProps}, ...childArray.map(asText))
}
