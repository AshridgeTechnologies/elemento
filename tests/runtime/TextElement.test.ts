import React from 'react'
import renderer from 'react-test-renderer'
import {TextElement} from '../../src/runtime/TextElement.js'

test('TextElement element produces output containing children', ()=> {
    const component = renderer.create(React.createElement(TextElement, null, 'Hello', 'where are you'))
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})