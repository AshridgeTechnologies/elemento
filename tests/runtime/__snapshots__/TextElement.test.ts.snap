// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextElement adds line break before line ends 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
      }
    }
    data-emotion="css d3wcwz-MuiTypography-root"
  />,
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
    id="page1.para1"
  >
    Hello
    <br />
    where are you
    <br />
    today?
    <br />
  </p>,
]
`;

exports[`TextElement element produces output containing ReactElement children 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
      }
    }
    data-emotion="css d3wcwz-MuiTypography-root"
  />,
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
    id="page1.para1"
  >
    Hello
    <p>
      where are you
    </p>
  </p>,
]
`;

exports[`TextElement element produces output containing string children 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
      }
    }
    data-emotion="css d3wcwz-MuiTypography-root"
  />,
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
    id="page1.para1"
  >
    Hello
    where are you
  </p>,
]
`;

exports[`TextElement element produces output containing value property of object 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-d3wcwz-MuiTypography-root{margin:0;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;margin-bottom:0.35em;}",
      }
    }
    data-emotion="css d3wcwz-MuiTypography-root"
  />,
  <p
    className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
    id="page1.para1"
  >
    where are you
  </p>,
]
`;
